mlflow/models/docker_utils.py

```diff
@@ -39,41 +40,43 @@
-    parsed_http_proxy = urlparse(http_proxy)
-    assert parsed_http_proxy.hostname is not None, "Invalid `http_proxy` hostname."
-    assert parsed_http_proxy.port is not None, f"Invalid proxy port: {parsed_http_proxy.port}"
-
-    parsed_https_proxy = urlparse(https_proxy)
-    assert parsed_https_proxy.hostname is not None, "Invalid `https_proxy` hostname."
-    assert parsed_https_proxy.port is not None, f"Invalid proxy port: {parsed_https_proxy.port}"
-
-    maven_proxy_options = (
-        "-DproxySet=true",
-        f"-Dhttp.proxyHost={parsed_http_proxy.hostname}",
-        f"-Dhttp.proxyPort={parsed_http_proxy.port}",
-        f"-Dhttps.proxyHost={parsed_https_proxy.hostname}",
-        f"-Dhttps.proxyPort={parsed_https_proxy.port}",
-        "-Dhttps.nonProxyHosts=repo.maven.apache.org",
-    )
+def get_maven_settings():
+    http_proxy = os.environ.get("http_proxy", os.environ.get("HTTP_PROXY"))
+    https_proxy = os.environ.get("https_proxy", os.environ.get("HTTPS_PROXY"))
+    no_proxy = os.environ.get("no_proxy", os.environ.get("NO_PROXY"))
 
-    if parsed_http_proxy.username is None or parsed_http_proxy.password is None:
-        return " ".join(maven_proxy_options)
+    non_proxy_hosts = "|".join(no_proxy.split(",")) if no_proxy is not None else None
 
-    return " ".join(
-        (
-            *maven_proxy_options,
-            f"-Dhttp.proxyUser={parsed_http_proxy.username}",
-            f"-Dhttp.proxyPassword={parsed_http_proxy.password}",
-        )
-    )
+    def to_maven_proxy(elem, name, url):
+        # Expects proxies as either PROTOCOL://{USER}:{PASSWORD}@HOSTNAME:PORT
+        # or PROTOCOL://HOSTNAME:PORT
+        parsed_url = urlparse(url)
+        assert parsed_url.hostname is not None, f"Invalid `{name}` hostname."
+        assert parsed_url.port is not None, f"Invalid proxy port: {parsed_url.port}"
+        ET.SubElement(elem, "id").text = name
+        ET.SubElement(elem, "active").text = "true"
+        ET.SubElement(elem, "protocol").text = parsed_url.scheme
+        ET.SubElement(elem, "host").text = parsed_url.hostname
+        ET.SubElement(elem, "port").text = str(parsed_url.port)
+        if parsed_url.username is not None:
+            ET.SubElement(elem, "username").text = parsed_url.username
+        if parsed_url.password is not None:
+            ET.SubElement(elem, "password").text = parsed_url.password
+        if non_proxy_hosts is not None:
+            ET.SubElement(elem, "nonProxyHosts").text = non_proxy_hosts
+
+    settings = ET.Element("settings")
+
+    proxies = ET.SubElement(settings, "proxies")
+
+    if http_proxy is not None:
+        http_proxy_settings = ET.SubElement(proxies, "proxy")
+        to_maven_proxy(http_proxy_settings, "http_proxy", http_proxy)
+    if https_proxy is not None:
+        https_proxy_settings = ET.SubElement(proxies, "proxy")
+        to_maven_proxy(https_proxy_settings, "https_proxy", https_proxy)
+
+    return ET.tostring(settings)
```

Comment:
Why we need to build `xml.etree.ElementTree` ?
it makes code complex.
Can we just generate `-Dhttp.proxyHost, -Dhttps.proxyHost, etc` configs like previous version code ?

Reply:
I agree that it adds complexity. Its unfortunate, but maven simply hasn't adopted modern practices when it comes to this. The alternative is to simply not support authenticated proxies for these two commands, which isn't great. If that's the case, then it should be explicitly documented, and the code above should fail before the Dockerfile is even generated if the proxy authentication is set with an error explaining this non-support.

Reply:
For context, its a common misconception, but the http(s).proxy(User/Password) properties are not actually part of the java standard (see https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/doc-files/net-properties.html). The Java standard library does not in fact implement authenticated proxies, at least not out of the box. Every Java tool must implement it itself, or use a third-party library that implements it for them. Many tools and libraries adopt the convention of using -Dhttp.proxyUser, etc, but this is not a requirement. Maven is an example of a tool which implements authenticated proxies its own way, and does not even know about such properties.